agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -62,17 +67,71 @@ def on_container_event(topic, e)
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
         @container_state_changed = true
+        handle_restart_on_die if e.status == 'die'.freeze
+      end
+    end
+
+    # Handles events when container has died
+    def handle_restart_on_die
+      unless @service_pod.running?
+        @restart_backoff_timer.cancel if @restart_backoff_timer
+        return
       end
+
+      return if apply_in_progress?
+
+      # cancel counters because we are called again
+      cancel_restart_timers
+
+      # backoff restarts
+      backoff = @restarts ** 2
+      backoff = max_restart_backoff if backoff > max_restart_backoff
+      if backoff == 0
+        info "restarting #{@service_pod.name_for_humans} because it has stopped"
+      else
+        info "restarting #{@service_pod.name_for_humans} because it has stopped (delay: #{backoff}s)"
+      end
+      @restart_backoff_timer = after(backoff) {
+        apply_restart if @container_state_changed
+      }
+      @restarts += 1
+    end
+
+    def max_restart_backoff
+      Kontena::Workers::ServicePodManager::LOOP_INTERVAL
+    end
+
+    def cancel_restart_timers
+      @restart_counter_reset_timer.cancel if @restart_counter_reset_timer
+      @restart_backoff_timer.cancel if @restart_backoff_timer
+    end
+
+    def apply_restart
```

Comment:
I don't see why there needs to be a separate `apply_restart`, why not include the `@restart_counter_reset_timer` -> `@restarts = 0` logic into the normal `apply`?

ATM if the `on_container_event` -> `handle_restart_on_die` schedules an `apply_restart`, but the `ServicePodManager` loop kicks in before the `@restart_backoff_timer` fires, then the `update` -> `apply` will actually restart the container, *without* resetting the `@restarts` counter. I'm not sure if the `@restart_backoff_timer` will actually end up resetting the `@restarts` counter in that case; it depends on what celluloid will run first after the exclusive apply block exits: does it run the `@restart_backoff_timer` with `@container_state_changed` still set to false, or does it first process the `on_container_event` with the `start` event that sets `@container_state_changed` back to true, allowing the `@restart_backoff_timer` to call `apply_restart` and reset the `@restarts` counter? I don't want to have to test it out...